### ISSUE TITLE
Feature request: task dependency graphs

### ISSUE THREAD

**Issue by dHannasch** [ray-project/ray#5958]
Ray allows us to implicitly compute according to any directed acyclic graph via having remote functions call other remote functions.
https://ray.readthedocs.io/en/latest/advanced.html#nested-remote-functions

However, consider a task that depends on *any* of several possible inputs. We do not know in advance which of the inputs will be ready soonest. We would like to run this task as soon as any of the inputs is ready.

(Compare to ipyparallel graph dependencies. https://ipyparallel.readthedocs.io/en/latest/task.html#graph-dependencies   The two types of dependencies, a node depending on *all* of its predecessors or a node depending on *any* of its predecessors, can ultimately encode any kind of dependency chain, though you might have to introduce artificial nodes into the dependency graph to encode more complicated situations.)

It would be useful to be able to tell Ray "this computation should be started as soon as it's ready to start, where ready-to-start is defined as follows..." based on other computation tasks, including tasks that themselves have not been started yet.

This isn't terribly important compared to the really key features of Ray --- obviously Ray running seamlessly and efficiently over all kinds of networks architectures is what really matters, more than the bells and whistles. But this is a situation that seems to come up a lot in scientific computing.

I've written up a proof of concept at https://github.com/dHannasch/task-dependency-graph. (It's probably horribly inefficient, because I'm still learning Ray. That, and I used NetworkX instead of a faster graph library for the first try.) The meat of it is all in https://github.com/dHannasch/task-dependency-graph/blob/master/src/task_dependency_graph/graph.py.

You can see an example illustrated at https://task-dependency-graph.readthedocs.io/en/latest/propagation.html.

There we have a simple grid graph where each node can be computed from *any* of its neighbors. (The first value must be computed by some external, presumed-expensive method that does not rely on having a neighbor value available.)

**Comment by ericl** [comment]
Can't you use ray.wait() to handle nondeterministic dependencies?

**Comment by robertnishihara** [comment]
Yeah, you can call `ray.wait` inside a function and pass in object IDs in a list. E.g.,

```python
import time
import numpy as np
import ray

ray.init()

@ray.remote
def f():
    time.sleep(np.random.uniform())
    print('Finished f', time.time())

@ray.remote
def wait_for_one_id(list_of_ids):
    # Return as soon as one of the IDs is ready.
    ray.wait(list_of_ids, num_returns=1)

@ray.remote
def g(wait_id):
    print('Started g', time.time())

# Start 3 f tasks.
ids = [f.remote() for _ in range(4)]
wait_id = wait_for_one_id.remote(ids)
# Start g as soon as the first f task finishes.
ray.get(g.remote(wait_id))
```

You should see an output like

```
(pid=45352) Finished f 1571709855.259553
(pid=45352) Started g 1571709855.261119
(pid=45350) Finished f 1571709855.68991
(pid=45330) Finished f 1571709856.002548
(pid=45351) Finished f 1571709856.159991
```

This isn't the nicest syntax, but it should be possible to come up with a helper function that takes care of some of the details. Let me know if this is doing what you're looking for.

**Comment by dHannasch** [comment]
Thank you!

You told me something I didn't know --- that ray.wait() can wait specifically for any of several inputs to be ready. When reading the Walkthrough, I saw "Return a list of IDs that are ready and a list of IDs that are not." and skipped past it thinking it would return immediately with whatever was done. But this looks very useful.
https://ray.readthedocs.io/en/latest/walkthrough.html#fetching-results
> If [timeout] is not set, the function simply waits until that number of objects is ready and returns that exact number of object IDs.

ray.wait() is definitely something that should be addressed in any proposal for task dependency graphs.

However, now that I go to address it, I don't see any way for ray.wait() to directly perform what I'm talking about.

(ray.wait() can definitely be *used* in the repo I linked --- I could make the "manager" Actor ray.wait() on all active workers in a loop, rather than give each worker a handle to call the manager back. I'm not sure whether that would be an improvement or not. I can't immediately think of any advantages or disadvantages. It's about the same amount of code either way.)

ray.wait() definitely works great when you have *exactly one* task that depends on any of some number of inputs --- just create workers for those tasks and call ray.wait() in your main Python script, as in the code you posted. But what about when you have lots of such tasks, as in the grid case?
https://task-dependency-graph.readthedocs.io/en/latest/propagation.html
When we have a grid where each node has two to four neighbors, we need to call ray.wait() once for each node, passing in its two to four neighbors. (Right?) Who's going to call ray.wait()? We'd have to immediately create an Actor for every single node in the graph and have each of them call ray.wait(). (Right?)

I might be missing something obvious, so I'm going to lay out my assumptions.

Possibly-wrong belief 1: **We should try not to create Actors (or Tasks, etc.) in significant numbers until they are ready to run.**

From reading the docs, it's not totally clear to me that Ray is even capable of reclaiming resources from an Actor (or Task) while it's ray.waiting. Even if Ray does do that, it sounds like Actors are fairly expensive to create and maintain.

https://ray.readthedocs.io/en/latest/walkthrough.html#remote-functions-tasks
> To specify a task’s CPU and GPU requirements, pass the num_cpus and num_gpus arguments into the remote decorator. The task will only run on a machine if there are enough CPU and GPU (and other custom) resources available to execute the task.
> If you do not specify any resources in the @ray.remote decorator, the default is 1 CPU resource and no other resources.
> If specifying CPUs, Ray does not enforce isolation (i.e., your task is expected to honor its request).
> The resource requirements of a task have implications for the Ray’s scheduling concurrency. In particular, the sum of the resource requirements of all of the concurrently executing tasks on a given node cannot exceed the node’s total resources.

If I am understanding this correctly, then once a remote function is created, it will reserve one CPU (or half a CPU, or two CPUs, etc) until it goes out of scope and gets garbage-collected.

If I am understanding this correctly, `num_cpus` CPUs are always reserved as long as a function is running. `num_cpus` defaults to 1, and must be set to *some* positive number for the function to run at all.

https://github.com/ray-project/ray/blob/master/python/ray/remote_function.py#L62
```
DEFAULT_REMOTE_FUNCTION_CPUS = 1
...
        self._num_cpus = (DEFAULT_REMOTE_FUNCTION_CPUS
                          if num_cpus is None else num_cpus)
```

Actors, it seems, are more complicated. The docs aren't totally clear on this point --- first the Walkthrough says a worker process is started as soon as an actor is instantiated, then the Actors page says the actor will acquire resources for its entire lifetime and if you do not specify CPU resources, the default is 1.
But then it goes on to say that "If you do not specify any resource requirements in the actor class’s remote decorator, then by default, the actor will not acquire any resources for its lifetime, but every time it executes a method, it will need to acquire 1 CPU resource." The source code seems to bear this out.

https://ray.readthedocs.io/en/latest/walkthrough.html#remote-classes-actors
```
When an actor is instantiated, the following events happen.
    A worker Python process is started on a node of the cluster.
```

https://ray.readthedocs.io/en/latest/actors.html#resources-with-actors
> If you specify resource requirements in an actor class’s remote decorator, then the actor will acquire those resources for its entire lifetime (if you do not specify CPU resources, the default is 1), even if it is not executing any methods. The actor will not acquire any additional resources when executing methods.
> If you do not specify any resource requirements in the actor class’s remote decorator, then by default, the actor will not acquire any resources for its lifetime, but every time it executes a method, it will need to acquire 1 CPU resource.
> Note that to create these actors successfully, Ray will need to be started with sufficient CPU resources

https://github.com/ray-project/ray/blob/master/python/ray/actor.py#L342
```
            cpus_to_use = ray_constants.DEFAULT_ACTOR_CREATION_CPU_SIMPLE
            actor_method_cpu = ray_constants.DEFAULT_ACTOR_METHOD_CPU_SIMPLE
```
https://github.com/ray-project/ray/blob/master/python/ray/ray_constants.py#L39
```
DEFAULT_ACTOR_METHOD_CPU_SIMPLE = 1
DEFAULT_ACTOR_CREATION_CPU_SIMPLE = 0
```

So...maybe...we can create an Actor without immediately claiming any resources for it. But if we call any method, it will claim resources until the method returns.

Possibly-wrong belief 2: **In order to make an Actor ray.wait(), we must call a method on it.**

Possibly-wrong belief 3: **In order to have dependency-graph-like logic without the full complicated data structure of the graph owner/manager, we must immediately create an Actor for every single node in the graph and call a method on each Actor to ray.wait() on its neighbors.**

The *graph* might not explicitly exist as a Python object, but we must have *something* to call ray.wait() for every single node in our (imaginary mathematical) graph. Right?

I apologize if this wasn't clear, but in general the number of tasks/nodes may far exceed the number of physical nodes in the hardware cluster. For example, the 64-grid might run on a cluster of only 8. (Which would be pretty much maximum efficiency in that case, since it would be very unlikely that there would ever be more than about 8 tasks ready to run in parallel.)

Even if we had access to a cluster big enough to reserve a physical node for every mathematical node, we wouldn't want to consume the entire cluster that way, because almost all the nodes would be sitting idle at any given time.

Thus,
Possibly-wrong belief 4: **In order to have dependency-graph-like logic without the full complicated data structure of the graph owner/manager, we must consume as many physical nodes in the cluster as we have nodes in the mathematical graph for the entire length of the computation.**

ray.wait() could certainly be used in the linked repo, to make the graph owner/manager wait() for all workers in a loop rather than wait for a callback:
```
while len(workers) > 0:
  answer = ray.wait([worker.getAnswer() for worker in workers], num_returns=1)
  # send answer to all successors in the graph
```
If that would make any significant difference compared to the graph owner/manager waiting for a callback to one of its methods, I'm failing to see it.
